server/src/main/java/io/unitycatalog/server/service/SchemaService.java

```diff
@@ -2,46 +2,164 @@
 public class SchemaService {
   private static final SchemaRepository SCHEMA_REPOSITORY = SchemaRepository.getInstance();
+  private static final CatalogRepository CATALOG_REPOSITORY = CatalogRepository.getInstance();
+  private final UnityCatalogAuthorizer authorizer;
+  private final UnityAccessEvaluator evaluator;
 
-  public SchemaService() {}
+  @SneakyThrows
+  public SchemaService(UnityCatalogAuthorizer authorizer) {
+    this.authorizer = authorizer;
+    evaluator = new UnityAccessEvaluator(authorizer);
+  }
 
   @Post("")
-  public HttpResponse createSchema(CreateSchema createSchema) {
-    return HttpResponse.ofJson(SCHEMA_REPOSITORY.createSchema(createSchema));
+  @AuthorizeExpression("""
+      #authorize(#principal, #metastore, OWNER) ||
+      #authorize(#principal, #catalog, OWNER) ||
+      #authorizeAll(#principal, #catalog, USE_CATALOG, CREATE_SCHEMA)
+      """)
+  @AuthorizeKey(METASTORE)
+  public HttpResponse createSchema(
+      @AuthorizeKey(value = CATALOG, key = "catalog_name") CreateSchema createSchema) {
+    SchemaInfo schemaInfo = SCHEMA_REPOSITORY.createSchema(createSchema);
+    createAuthorizations(schemaInfo);
+    return HttpResponse.ofJson(schemaInfo);
   }
 
   @Get("")
+  @AuthorizeExpression("#defer")
   public HttpResponse listSchemas(
       @Param("catalog_name") String catalogName,
       @Param("max_results") Optional<Integer> maxResults,
       @Param("page_token") Optional<String> pageToken) {
-    return HttpResponse.ofJson(SCHEMA_REPOSITORY.listSchemas(catalogName, maxResults, pageToken));
+    ListSchemasResponse listSchemasResponse =
+        SCHEMA_REPOSITORY.listSchemas(catalogName, maxResults, pageToken);
+    filterSchemas("""
+        #authorize(#principal, #metastore, OWNER) ||
+        #authorize(#principal, #catalog, OWNER) ||
+        #authorize(#principal, #schema, OWNER) ||
+        (#authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG))
+        """,
+        listSchemasResponse.getSchemas());
+    return HttpResponse.ofJson(listSchemasResponse);
   }
 
   @Get("/{full_name}")
-  public HttpResponse getSchema(@Param("full_name") String fullName) {
+  @AuthorizeExpression("""
+      #authorize(#principal, #metastore, OWNER) ||
+      #authorize(#principal, #schema, OWNER) ||
+      (#authorize(#principal, #schema, USE_SCHEMA) && #authorize(#principal, #catalog, USE_CATALOG))
+      """)
+  @AuthorizeKey(METASTORE)
+  public HttpResponse getSchema(@Param("full_name") @AuthorizeKey(SCHEMA) String fullName) {
     return HttpResponse.ofJson(SCHEMA_REPOSITORY.getSchema(fullName));
   }
 
   @Patch("/{full_name}")
-  public HttpResponse updateSchema(@Param("full_name") String fullName, UpdateSchema updateSchema) {
+  @AuthorizeExpression("""
+      #authorize(#principal, #metastore, OWNER) ||
+      #authorize(#principal, #schema, OWNER) ||
+      #authorizeAll(#principal, #catalog, USE_CATALOG, USE_SCHEMA) ||
```

Comment:
Does this mean a principal with USE_CATALOG and USE_SCHEMA on the catalog can update the schema?

Reply:
This looks like the entity should probably be `#schema` where since permissions are inherited it would essentially be checking `USE_CATALOG` on the catalog and `USE_SCHEMA` on the schema.